Collect binaries when the JVM crashes during a Jenkins build

jdmpapin opened the pull request:
Binaries help in analyzing the core file to debug the crash.

jdmpapin commented:
A few questions:
- Are any of the other (already) collected files useful only for analyzing cores? e.g. maybe DDR things? This change already makes inclusion of `*.dSYM` conditional on the existence of a core.
- Do we also want to be able to run the build as @pshipton [suggested](https://github.com/eclipse-openj9/openj9/issues/16400#issuecomment-1338513580)? Re-running likely requires more than just the executables. For example, it probably also requires jars (at least). It's unclear to me how we would predict the minimal subset of files that allow running the program, so if we want to be able to run it, I think we'd need to pick up e.g. all of `build/*/jdk/*` and `build/*/images/{jdk,j2sdk-image}/*` (in which case we could probably stop worrying about particular file extensions for the binaries and debug info).
- How do we test this? Perhaps by deliberately failing some PR builds?

pshipton commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -529,10 +539,38 @@ def archive_diagnostics() {
+        archiveCmd = "pax -wzf ${diagnosticsFilename}"
     } else {
-        sh "find . -name 'core.*.dmp' -o -name 'javacore.*.txt' -o -name 'Snap.*.trc' -o -name 'jitdump.*.dmp' -o -name ddr_info -o -name j9ddr.dat -o -name superset.dat -o -name '*.annt.h' -o -name '*.stub.h' -o -name '*.dSYM' -o -path '*/make-support/failure-logs/*' | sed 's#^./##' | tar -zcvf ${diagnosticsFilename} -T -"
+        findCmd += " -o -name ddr_info"
```
These things are also relevant to z/OS. Not sure we didn't add them when collecting diagnostics on z/OS, probably a mistake.

jdmpapin replied:
These are now added regardless of OS

pshipton commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -529,10 +539,38 @@ def archive_diagnostics() {
+        // JDK bin directory. On Windows this will contain .exe and .dll files,
+        // but on most systems the binaries in it won't have any file extension
+        // we could use to identify them, so just include everything in there.
+        findCmd += " -o -path '*/build/*/jdk/bin/*'"
```
I'd collect everything here no matter if the core is found or not. Maybe they can be used to recreate the failure manually.

jdmpapin replied:
Ah, I was just trying to avoid collecting everything when we have the usual kind of build failure, i.e. an error where the build compiler complains about the source, or javac complains about some Java source. Failures like that should be possible to recreate manually just based on the source.

If you want to collect this all even in those cases, I can easily make that change.

pshipton replied:
It's a good idea. I was thinking about the case where the core file didn't work. How about collecting the info if any core, javacore, Snap, jitdump is present.

jdmpapin replied:
Now collecting this type of stuff if any JVM crash data is found

pshipton commented:
> Do we also want to be able to run the build

We want to be able to diagnose the problem. Anything we can collect to help with what is relevant. I'm not against collecting the class libraries in the diagnostics. Every successful build archives the libraries and binaries. Doing the wildcards seems fine. It's not very useful when a failure occurs and we can't do anything about it.

> How do we test this? Perhaps by deliberately failing some PR builds?

I expect so.

pshipton commented:
@keithc-ca @AdamBrousseau fyi

pshipton commented:
> I'll ignore the file types and go by directory

Chatting with Keith, with the wildcards we should exclude some things we don't need to reduce the size. The most obvious being .o and .obj files.

jdmpapin commented:
Ah, I thought the object files weren't in the directories I was suggesting, at least based on my own build directories on my laptop. It still looks that way for 11 and 17, but I see that 8 does put object files within `build/$platform/jdk`

keithc-ca commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -500,6 +500,16 @@ def fetchFile(src, dest) {
+
+    def archiveCmd = "tar -zcvf ${diagnosticsFilename} -T -"
+
+    def findNameCore = "-name 'core.*.dmp'"
+    def findCmd = "find ."
+    findCmd += " ${findNameCore}"
```
Suggest inserting "find ." into `findNameCore`:
```sh
    def findNameCore = "find . -name 'core.*.dmp'"
    def findCmd = "${findNameCore}"
```
and fixing the other use of `${findNameCore}`.

jdmpapin replied:
Done - this is now `findCrashDataCmd`

keithc-ca commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -500,6 +500,16 @@ def fetchFile(src, dest) {
+    def findNameCore = "-name 'core.*.dmp'"
+    def findCmd = "find ."
+    findCmd += " ${findNameCore}"
+    findCmd += " -o -name 'javacore.*.txt'"
+    findCmd += " -o -name 'Snap.*.trc'"
+    findCmd += " -o -name 'jitdump.*.dmp'"
```
Can you please sort the additions alphabetically, here and below?

jdmpapin replied:
Sorted these and the debug info globs. I think the lines adding DDR-related stuff are already sorted well enough

jdmpapin commented:
OK, so to check which directories contain binaries that we actually run while building, I built Java 8, 11, 17, and 19 on my laptop while logging all processes with `strace -f -e trace='%process'`. In describing the results, I'll call the particular platform-specific output directory `$b`, so e.g. `$b` might be `build/linux-x86_64-server-release`.

For the Java 8 build, the only Java executable within build/ that it ran was `$b/j2re-image/bin/java -version` at the end, presumably to make sure that it succeeds. For everything else it used the boot JDK.

The Java 11, 17, and 19 builds ran Java executables from `$b/jdk/`, and from nowhere else within build/. In particular, they ran the following executables:
- `$b/jdk/bin/java`
- `$b/jdk/bin/jmod`
- `$b/jdk/bin/jlink`

Apparently none of the builds needed to run their own `javac`. The versions that used to need to do so can now rely on a new enough boot JDK. Later versions introducing new Java language features and using them in the JCL will need to do so once again, at least initially, and I can't necessarily predict which `javac` they will run, but my guess is that it will be `$b/jdk/bin/javac`.

In all of the builds, there were some additional executables within build/ that were run - more specifically they were within `$b/vm/runtime`:
- `constgen`
- `omr_ddrgen`
- `tracegen`
- `tracemerge`

I don't think these run JVMs, but I suppose it's possible regardless that they could crash. Do we want to try to notice crashes from these and/or include these binaries? My `vm/runtime` directory is huge - I think it contains most of the object files - so we'd have to pick them up in a targeted way. My copies of these executables are linked only against system libraries and (in the case of `constgen` and `omr_ddrgen`) against `$b/vm/runtime/lib{omrsig,j9thr29}.so`. Noticing crashes from these may be harder since they might not produce core files with the predictable naming pattern used by the JVM, or they might by default not produce a core (e.g. on Windows?).

Finally, Java 8 also ran something called `$b/jdk/gensrc_x11wrappers/sizer.64.exe`.

----

My suggestion is that for the moment I think it should suffice to pick up `$b/jdk/` (excluding object files).

pshipton commented:
> My suggestion is that for the moment I think it should suffice to pick up $b/jdk/ (excluding object files)

That works for me. Having a problem with the other executables isn't common, and likely if they have a problem it's easily repeatable and can be recreated manually. 

I suspect javac isn't used because the compiler can be run by running java.

jdmpapin commented:
Updated based on my previous comment and to address review feedback

jdmpapin commented:
I don't believe this revision should include any object files, even though it doesn't explicitly exclude them. We can confirm in testing

jdmpapin commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -529,10 +538,50 @@ def archive_diagnostics() {
     }
+
+    def findDiagnosticsCmd = findCrashDataCmd
+    findDiagnosticsCmd += " -o -type d -name ddr_info -prune"
+    findDiagnosticsCmd += " -o -name j9ddr.dat"
+    findDiagnosticsCmd += " -o -name superset.dat"
+    findDiagnosticsCmd += " -o -name '*.annt.h'"
+    findDiagnosticsCmd += " -o -name '*.stub.h'"
+    findDiagnosticsCmd += " -o -path '*/make-support/failure-logs/*'"
```
Are any of these useful only for debugging JVM crashes? I imagine it's the case for DDR stuff (including, I assume, `ddr_info` and `j9ddr.dat`), but can somebody confirm? And I'm not sure what the rest of these things are

keithc-ca replied:
I'm not sure there's much value in capturing `j9ddr.dat`, `superset.dat` or the `.annt` or `.stub` files.
`ddr_info` is being pruned here so it won't be include (we need to be mindful of operator precedence) with that fragment.
OpenJDK makefiles capture command lines, etc. in `failure-logs` so those might be useful (they're relatively small files).

jdmpapin replied:
With `-prune`, `find` will still find the `ddr_info` directory. It just won't descend into it. Compare:
```
$ find openj9-openjdk-jdk8/build/ -type d -name ddr_info -o -name j9jvmti.txt
openj9-openjdk-jdk8/build/linux-x86_64-normal-server-release/vm/ddr_info
openj9-openjdk-jdk8/build/linux-x86_64-normal-server-release/vm/ddr_info/targets/j9jvmti.txt
```
vs.
```
$ find openj9-openjdk-jdk8/build/ -type d -name ddr_info -prune -o -name j9jvmti.txt
openj9-openjdk-jdk8/build/linux-x86_64-normal-server-release/vm/ddr_info
```
Descending into the directory is pointless since the directory itself is already listed to be included in the archive, and it will be included recursively. For example, take a look at [this crash](https://openj9-jenkins.osuosl.org/job/Build_JDK19_x86-64_mac_Nightly/127/consoleText), where the files being added to the archive beneath `ddr_info` do not match any of the other patterns that the `find` command was looking for:
```
[2022-12-01T01:08:36.763Z] build/macosx-x86_64-server-release/vm/ddr_info/
[2022-12-01T01:08:36.763Z] build/macosx-x86_64-server-release/vm/ddr_info/targets/
[2022-12-01T01:08:36.763Z] build/macosx-x86_64-server-release/vm/ddr_info/targets/ddrgentest.txt
[2022-12-01T01:08:36.763Z] build/macosx-x86_64-server-release/vm/ddr_info/targets/omrport_obj.txt
[2022-12-01T01:08:36.763Z] build/macosx-x86_64-server-release/vm/ddr_info/targets/j9jvmti.txt
...
```
So that's why I put `-prune` there, and also on the other directories that I want to include in their entirety.

However, I haven't checked yet to make sure that `pax -wzf ...` will also include directories recursively. If it doesn't, then I'll have to replace `-type d -path '...' -prune` with `-path '.../*'`.

jdmpapin replied:
They're originally from https://github.com/eclipse-openj9/openj9/pull/9670

jdmpapin replied:
Alright, so I guess out of these lines that leaves only `ddr_info` and `make-support/failure-logs`.

@keithc-ca, should we keep collecting `ddr_info`? Based on your earlier comment, I'm not totally clear on your position on that

jdmpapin replied:
OK, well, I've removed the others and kept `ddr_info` for now. It's easy to remove it later if we want to

jdmpapin commented:
Removed some clauses that Keith suggested are not useful

jdmpapin commented:
As I suggested at the beginning of the comments, I'll try to test these changes by injecting bugs causing build failures into this PR branch, and then observing the behaviour of PR builds. In preparation, I've converted this PR to a draft to ensure that we don't accidentally merge any deliberately injected bug.

jdmpapin commented:
Any further comments before I start testing?

pshipton commented:
Not from me.

jdmpapin commented:
Jenkins compile xlinux jdk17

jdmpapin commented:
Jenkins compile win jdk8

jdmpapin commented:
From the x86-64 Linux Java 17 crash, I can run `jdk/bin/java -Xint -version`, and I can see the crash by removing `-Xint`. The crashed process command line and output are in make-support/failure-logs/, which would be good to help reproduce an actual failure. With `set sysroot ...` I see symbols and line numbers in GDB, and by checking out the commits from `-version` and specifying `dir ...`, GDB also prints the source line. The compressed archive is 483MB. It doesn't include any object files

Testing Java 8 on Windows now

jdmpapin commented:
No diagnostics were collected for the Windows Java 8 failure. It seems that the crash doesn't prevent `make` from succeeding:
```
[2022-12-13T18:24:27.409Z] Assertion failed at F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\openj9\runtime\compiler\ilgen\Walker.cpp:189: false
[2022-12-13T18:24:27.409Z] VMState: 0x00050080
[2022-12-13T18:24:27.409Z] 	oh no, the JVM crashed!
[2022-12-13T18:24:27.409Z] compiling java/lang/String.lengthInternal()I at level: warm
[2022-12-13T18:24:27.409Z] 
[2022-12-13T18:24:27.409Z] JVMDUMP052I JIT dump recursive crash occurred on diagnostic thread
[2022-12-13T18:24:27.409Z] JVMDUMP010I JIT dump written to F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\jitdump.20221213.122422.4716.0004.dmp
[2022-12-13T18:24:27.410Z] JVMDUMP013I Processed dump event "gpf", detail "".
[2022-12-13T18:24:27.972Z] ----- Build times -------
[2022-12-13T18:24:27.972Z] Start 2022-12-13 11:29:58
[2022-12-13T18:24:27.972Z] End   2022-12-13 12:24:28
[2022-12-13T18:24:27.972Z] 00:00:37 corba
[2022-12-13T18:24:27.972Z] 00:00:41 demos
[2022-12-13T18:24:27.972Z] 00:03:15 docs
[2022-12-13T18:24:27.972Z] 00:06:39 images
[2022-12-13T18:24:27.972Z] 00:00:35 jaxp
[2022-12-13T18:24:27.972Z] 00:03:24 jaxws
[2022-12-13T18:24:27.972Z] 00:08:20 jdk
[2022-12-13T18:24:27.972Z] 00:01:07 langtools
[2022-12-13T18:24:27.972Z] 00:00:21 nashorn
[2022-12-13T18:24:27.972Z] 00:54:30 TOTAL
[2022-12-13T18:24:27.972Z] -------------------------
[2022-12-13T18:24:27.972Z] Finished building OpenJDK for target 'all'
```
Then the build failed when `java -version` crashed [here](https://github.com/eclipse-openj9/openj9/blob/108f91867785b9e961ed8b23283f0b149c112311/buildenv/jenkins/common/build.groovy#L311-L315).

jdmpapin commented:
I've opened ibmruntimes/openj9-openjdk-jdk8#639 for cases where `java -version` fails during `make` for a Java 8 build. Particularly, the failure should now cause `make` to fail as well, which in the context of this PR will cause us to run `archive_diagnostics()`.

keithc-ca commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -529,10 +538,46 @@ def archive_diagnostics() {
+            // that the binaries there can't be run from that location). But it
+            // does run images/j2re-image/bin/java -version, so instead of jdk/
+            // get images/j2re-image/.
+            findDiagnosticsCmd += " -o ${isJ2reImageDir}"
```
Perhaps this needs `-prune` like other things we expect to be directories?

jdmpapin commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -529,10 +538,46 @@ def archive_diagnostics() {
+            // The separate debug info is in vm/runtime/, not images/j2re-image/.
+            def vmRtGlob = "${tgtGlob}/vm/runtime"
+            findDiagnosticsCmd += " -o -path '${vmRtGlob}/*.debuginfo'"
+            findDiagnosticsCmd += " -o -path '${vmRtGlob}/*.dSYM'"
```
Oh, rather, it looks like the directories I'm having it slurp up will already contain all of the separate debug info, so I think I can stop even trying to match this stuff by file extension

jdmpapin replied:
I thought there was no separate debug info in images/j2re-image/, but it is there. It's just zipped and using a different file extension

jdmpapin commented:
I've updated build.groovy based on some new observations:
- With `RELEASE` we know the exact path to the build output directory, so we don't need to glob to find it. Replace `tgtGlob` with a non-globbing `tgt`.
- Stop looking for Java 8 separate debug info beneath vm/runtime/. [It should already be in j2re-image](https://github.com/eclipse-openj9/openj9/pull/16426#discussion_r1086811897).
- There is a final `java -version` that runs strictly after `make` in `stage('Java Version')`. Failures there would fail the build, but they wouldn't cause it to collect diagnostic data. Such failures will now trigger `archive_diagnostics`, and if crash data is detected, we'll also collect the binaries from the directory used to run `java -version` (which is different from the one used to run `java` during `make`).
- Specify the path to `make-support` instead of globbing. It's directly beneath `tgt`.
- Match `make-support/failure-logs` as a directory instead of globbing for each file within it.

I also rebased at the same time, so the force-push diff is very large - sorry - but thankfully build.groovy seems to be among the files shown.

jdmpapin commented:
Jenkins compile xlinux jdk17

jdmpapin commented:
Jenkins compile win jdk8

jdmpapin commented:
Fixed a syntax error I introduced in the previous push

jdmpapin commented:
Jenkins compile xlinux jdk17

jdmpapin commented:
Jenkins compile win jdk8

jdmpapin commented:
The x86-64 Linux Java 17 crash collected all the files I expect into the diagnostics tarball, in particular:
- the crash data itself (core, javacore, jitdump, Snap)
- build/linux-x86_64-server-release/jdk/ (recursively)
- build/linux-x86_64-server-release/support/modules_libs/ (recursively)
- build/linux-x86_64-server-release/make-support/failure-logs/ (recursively)
- build/linux-x86_64-server-release/vm/ddr_info/ (recursively)

My [previous x86-64 Linux Java 17 crash test](https://github.com/eclipse-openj9/openj9/pull/16426#issuecomment-1349186008) shows that those files are sufficient to run the `java` command (at least `java -version`...) and reproduce the problem, and to get the libraries loaded into GDB with debug info.

----

The x86-64 Windows Java 8 crash collected the following, also as expected:
- the crash data itself
- build/windows-x86_64-normal-server-release/images/j2re-image/ (recursively)
- build/windows-x86_64-normal-server-release/vm/ddr_info/ (recursively)

(Neither support/modules_libs/ nor make-support/failure-logs/ exist in Java 8 builds.)

I downloaded the diagnostics file and I was able to reproduce the crash in `java -version`. The crash disappears if I pass `-Xint` or `set TR_dontCrash=1`, and I can successfully get the version output. PDB files are present, and I'm able to load symbols in WinDBG.

jdmpapin commented:
Jenkins compile amac jdk11

jdmpapin commented:
Jenkins compile aix jdk19

jdmpapin commented:
In the AIX crash, the diagnostics tarball included only make-support/failure-logs/ and vm/ddr_info/. It seems that somehow the options to include the actual JDK were not passed to `find`. And the options meant to match the crash data *were* passed to `find`, but no crash data made it into the archive, even though the first pass of `find` seems to have found some. I'm not sure how any of this is possible :confused:
```
[Pipeline] sh
+ find . -name core.*.dmp -o -name javacore.*.txt -o -name jitdump.*.dmp -o -name Snap.*.trc
+ grep -q .
[Pipeline] sh
+ find . -name core.*.dmp -o -name javacore.*.txt -o -name jitdump.*.dmp -o -name Snap.*.trc -o -type d -name ddr_info -prune -o -type d -path ./build/aix-ppc64-server-release/make-support/failure-logs -prune
+ sed s#^./##
+ tar -zcvf Build_JDK19_ppc64_aix_Personal-69-20230127-230921-diagnostics.tar.gz -T -
build/aix-ppc64-server-release/make-support/failure-logs/
build/aix-ppc64-server-release/make-support/failure-logs/jdk__optimize_image_exec.cmdline
build/aix-ppc64-server-release/make-support/failure-logs/jdk__optimize_image_exec.log
build/aix-ppc64-server-release/vm/ddr_info/
build/aix-ppc64-server-release/vm/ddr_info/sets/
build/aix-ppc64-server-release/vm/ddr_info/sets/...
build/aix-ppc64-server-release/vm/ddr_info/targets/
build/aix-ppc64-server-release/vm/ddr_info/targets/...
```

keithc-ca commented:
> I'm not sure how any of this is possible

Sometimes, commands are issued from the `$(TOPDIR)/make` directory even though they may switch directories leaving those artifacts elsewhere. Perhaps `find` should be explicitly told to start looking in `$(TOPDIR)` or `$(OUTPUTDIR)` instead of `.`.

jdmpapin commented:
From the paths that were picked up by `tar`, and from the fact that `-path ./build/aix-ppc64-server-release/make-support/failure-logs` matched something, I think it's safe to conclude that `find` is already running in `$(TOPDIR)`

jdmpapin commented:
Retrying AIX to see if it was some kind of fluke

Jenkins compile aix jdk19

jdmpapin commented:
In the Mac failure, I was able to run the JVM and either repro the crash or print the version successfully. I tried to check that we can load symbols for the core in the debugger, but LLDB fails to even identify the libraries. I don't know what's going on, but this is the second time LLDB on Mac has flaked out on me. Last time I was able to analyze the core on a different machine. This time, I've tried two machines with the same results on both. So to check on the symbols I'm retrying Mac as well, but on x86. I very much doubt that the ISA will affect the set of files that need to be included in the tarball in order to get symbols. I'm hoping that it will be easier for me to find an x86 Mac with an LLDB that works on the core.

Jenkins compile xmac jdk11

jdmpapin commented:
OK, I think the AIX problem is actually that no crash data was found. Reading my changes again, the files collected and the second `find` command are both consistent with that. So I'm pretty sure I was just mistaken. Additionally, the log shows this:
```
Unhandled exception in signal handler. Protected function: generateDiagnosticFiles (0x0)
```
It's possible that I'm causing too many concurrent crashes. I'll revise my test commit to crash only once and try again.

jdmpapin commented:
Updated test commit to crash only once. We'll see if `std::atomic<bool>` is available in our AIX PR build...

I also noticed that because we know the exact path to the images/ directory, we don't need `find` to detect the presence of images/j2re-image/. It's possible to just `test -d` instead. So I've updated the script to do that.

jdmpapin commented:
Jenkins compile aix jdk19

jdmpapin commented:
Making sure that JDK8 is still detected properly after my most recent change.

Jenkins compile xlinux jdk8

jdmpapin commented:
No luck with `std::atomic`. I'll go find a VM function to call instead.
```
The ATM_RMW      operation is not supported by the target hardware.
```

jdmpapin commented:
OK, updated my test crash-once logic to use `VM_AtomicSupport` instead of `std::atomic` so that it should work on AIX. Hopefully this will allow the AIX build crash to produce crash data and actually exercise my changes.

It should still be fine to look at [my previous x86-64 Mac JDK11 build](https://openj9-jenkins.osuosl.org/job/Build_JDK11_x86-64_mac_Personal/455/). The only changes since then are in the test PR and in distinguishing JDK8 from 11+. The binaries collected in this one are good. I *think* the symbols (`*.dSYM`) are good too - LLDB claims to have loaded them - but I'm continuing to have other problems with LLDB so it's hard to be sure. I've asked somebody for some help in checking that the symbols are in fact present.

[My previous x86-64 Linux JDK8 build](https://openj9-jenkins.osuosl.org/job/Build_JDK8_x86-64_linux_Personal/345/) OTOH looks like it mysteriously got stuck. I've cancelled it and I'll launch a new one to make sure that we can still properly recognize JDK8 and collect images/j2re-image instead of jdk/.

jdmpapin commented:
Jenkins compile aix jdk19

jdmpapin commented:
Jenkins compile xlinux jdk8

jdmpapin commented:
Jenkins compile aix jdk19

jdmpapin commented:
Jenkins compile xlinux jdk8

jdmpapin commented:
@nbhuiyan got the stack trace from the Mac crash (thanks!), and the symbols in the diagnostics tarball are good.

jdmpapin commented:
The AIX build failed to generate any crash data again:
```
Unhandled exception in signal handler. Protected function: generateDiagnosticFiles (0x0)
```
It really should generate crash data, but that's a separate issue. I think for now this PR will have to proceed without verifying that the collected binaries and debug info are sufficient on AIX :shrug:

I was also going to use the AIX failure as the JDK11+ case to verify that the right branch is taken (to contrast with the JDK8 failure) after the `test -d` change. Now it seems that won't work, so I'll launch another build just for that purpose. This one should still collect jdk/ and support/modules_libs/ instead of trying to collect images/j2re-image/.

Jenkins compile xlinux jdk11

jdmpapin commented:
OK, the `test -d` change is working. The JDK11 crash still picked up jdk/ and support/modules_libs/, and the JDK8 crash picked up images/j2re-image/ instead, as expected.

jdmpapin commented:
Removed the JIT crash and instead injected an error into Class.java to test specifically that it doesn't cause us to collect binaries.

Jenkins compile xlinux jdk17

jdmpapin commented:
No binaries collected, as expected

jdmpapin commented:
I think this should be good to go now

jdmpapin commented:
@pshipton could you please review again?

keithc-ca commented on buildenv/jenkins/common/build.groovy:
```diff
@@ -310,7 +310,15 @@ def build() {
     stage('Java Version') {
         dir(OPENJDK_CLONE_DIR) {
-            sh "build/$RELEASE/images/$JDK_FOLDER/bin/java -Xjit -version"
+            // Include the leading ./ here so that jdkImageDir is suitable for
+            // use in find . -path '${jdkImageDir}'.
```
Please explain this change. I don't see how this relates to use of the `find` utility.

jdmpapin replied:
When we run `find . [..]`, the path to each file considered (other than `.` itself) will start with `./`. This is the path that we're checking with `-path`. So e.g. `find . -path build` will never match anything. Instead it to match the build directory it would be necessary to specify `-path ./build`.

keithc-ca replied:
If the comment made reference to the use of `jdkImageDir` in the call to `archive_diagnostics()` it might be clearer.

This seems a little fragile: `archive_diagnostics()` will not accept any path. Maybe `archive_diagnostics()` should be adding the `./` prefix rather than rely on the caller to do so.

jdmpapin replied:
I've removed the leading `./` here and the comment explaining it, and I've made `archive_diagnostics()` add it itself (when missing).

jdmpapin commented:
Updated based on Keith's latest review comment
